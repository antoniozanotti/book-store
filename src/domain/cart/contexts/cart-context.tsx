"use client";

import {
  createContext,
  useContext,
  useReducer,
  ReactNode,
  useEffect,
} from "react";
import { CartType } from "../types/cart-type";
import { CartContextType } from "../types/cart-context-type";
import { useLocalStorage } from "usehooks-ts";
import { useProductsByIdsQuery } from "@/domain/product/queries/use-products-by-ids-query";
import { cartReducer } from "../reducers/cart-reducer";

const initialState: CartType = { items: [] };

const CartContext = createContext<CartContextType>({
  state: initialState,
  dispatch: () => {},
  getCartItemsQty: () => {},
  cartItemsSubtotal: () => {},
});

export function CartProvider({ children }: { children: ReactNode }) {
  const [localState, setLocalState] = useLocalStorage<CartType>(
    "cart",
    initialState
  );
  const [state, dispatch] = useReducer(cartReducer, localState);
  const cartItems = state.items;
  const responses = useProductsByIdsQuery(cartItems.map((item) => item.id));

  function getCartItemsQty(): number {
    const cartItems = state.items;
    return cartItems.reduce((acc, cur) => {
      return acc + cur.quantity;
    }, 0);
  }

  function cartItemsSubtotal(): number {
    let subTotal = 0;
    cartItems.forEach((item) => {
      responses.some((response) => {
        const product = response.data;
        if (!product) {
          return true;
        }
        if (item.id == product.id) {
          subTotal += parseInt(product.price.toString()) * item.quantity;
          return true;
        }
      });
    });
    return subTotal;
  }

  useEffect(() => {
    setLocalState(state);
  }, [state, setLocalState]);

  return (
    <CartContext.Provider
      value={{ state, dispatch, getCartItemsQty, cartItemsSubtotal }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCartContext(): CartContextType {
  return useContext(CartContext);
}
