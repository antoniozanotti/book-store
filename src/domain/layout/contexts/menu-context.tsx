"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { MenuType } from "../types/menu-type";
import { MenuContextType } from "../types/menu-context-type";
import { menuReducer } from "../reducers/menu-reducer";

const initialState: MenuType = { isOpen: false };

const MenuContext = createContext<MenuContextType>({
  state: initialState,
  dispatch: () => {},
});

export function MenuProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(menuReducer, initialState);

  return (
    <MenuContext.Provider value={{ state, dispatch }}>
      {children}
    </MenuContext.Provider>
  );
}

export function useMenuContext(): MenuContextType {
  return useContext(MenuContext);
}
