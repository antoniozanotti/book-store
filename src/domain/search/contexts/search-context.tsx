"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { SearchContextType } from "../types/search-context-type";
import { useSearchParams } from "next/navigation";
import { getSearchParams } from "../services/search-service";
import { searchReducer } from "../reducers/search-reducer";

const SearchContext = createContext<SearchContextType>({} as SearchContextType);

export function SearchProvider({ children }: { children: ReactNode }) {
  const searchParams = useSearchParams();
  const { keyword, categories, minPrice, maxPrice, minYear, maxYear, orderBy } =
    getSearchParams(searchParams);
  const initialStateBySearchParams = {
    keyword: keyword,
    categories: categories,
    minprice: minPrice,
    maxprice: maxPrice,
    minyear: minYear,
    maxyear: maxYear,
    orderby: orderBy,
  };
  const [state, dispatch] = useReducer(
    searchReducer,
    initialStateBySearchParams
  );

  return (
    <SearchContext.Provider value={{ state, dispatch }}>
      {children}
    </SearchContext.Provider>
  );
}

export function useSearchContext(): SearchContextType {
  return useContext(SearchContext);
}
