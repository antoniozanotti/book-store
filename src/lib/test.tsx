import React, { ReactElement, ReactNode } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <div data-testid="container">{children}</div>;
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";
import { usePathname } from "next/navigation";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
    notFound: () => null,
    usePathname: () => null,
    useSearchParams: () => null,
  };
});
