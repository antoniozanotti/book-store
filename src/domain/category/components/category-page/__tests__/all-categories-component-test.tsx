import { render, screen } from "@/lib/test";
import { AllCategoriesComponent } from "../all-categories-component";
import { LayoutProvider } from "@/domain/layout/contexts/layout-context";
import { categoriesWithPageMock } from "@/domain/category/__tests__/categories-with-page-mock";
import { CategoryPageProvider } from "@/domain/category/contexts/category-page-context";
import { categoryWithPageAndProductsMock } from "@/domain/category/__tests__/category-with-page-and-products-mock";

describe("AllCategoriesComponent", () => {
  it("should render all categories", async () => {
    const { container } = render(
      <LayoutProvider categories={categoriesWithPageMock}>
        <AllCategoriesComponent />
      </LayoutProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render all categories and bold current category", async () => {
    const { container } = render(
      <LayoutProvider categories={categoriesWithPageMock}>
        <CategoryPageProvider category={JSON.parse(JSON.stringify(categoryWithPageAndProductsMock))}>
          <AllCategoriesComponent />
        </CategoryPageProvider>
      </LayoutProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
