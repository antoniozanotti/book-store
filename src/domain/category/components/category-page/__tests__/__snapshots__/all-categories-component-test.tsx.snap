// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllCategoriesComponent should render all categories 1`] = `
<div>
  <div
    data-testid="container"
  >
    <nav
      aria-label="Categories"
      class="p-10"
    >
      <h2
        class="font-primary text-pretty text-2xl"
      >
        Categories
      </h2>
      <ul
        class="text-sm flex flex-col divide-y"
      >
        <li
          class="py-2 "
        >
          <a
            href="/category/arts-and-entertainment"
            title="Arts and Entertainment"
          >
            Arts and Entertainment
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/business-and-technology"
            title="Business and Technology"
          >
            Business and Technology
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/graphic-novels"
            title="Graphic Novels"
          >
            Graphic Novels
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/history-biography-and-social-science"
            title="History, Biography, and Social Science"
          >
            History, Biography, and Social Science
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/home-and-lifestyle"
            title="Home and Lifestyle"
          >
            Home and Lifestyle
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/horror"
            title="Horror"
          >
            Horror
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/languages-and-travel"
            title="Languages and Travel"
          >
            Languages and Travel
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/lgbtqia"
            title="LGBTQIA+"
          >
            LGBTQIA+
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/literature-and-poetry"
            title="Literature and Poetry"
          >
            Literature and Poetry
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/mystery"
            title="Mystery"
          >
            Mystery
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/romance"
            title="Romance"
          >
            Romance
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/science-and-math"
            title="Science and Math"
          >
            Science and Math
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/science-fiction-and-fantasy"
            title="Science Fiction and Fantasy"
          >
            Science Fiction and Fantasy
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/spirituality-and-philosophy"
            title="Spirituality and Philosophy"
          >
            Spirituality and Philosophy
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/wellness-sports-and-outdoors"
            title="Wellness, Sports, and Outdoors"
          >
            Wellness, Sports, and Outdoors
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/young-adult"
            title="Young Adult"
          >
            Young Adult
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`AllCategoriesComponent should render all categories and bold current category 1`] = `
<div>
  <div
    data-testid="container"
  >
    <nav
      aria-label="Categories"
      class="p-10"
    >
      <h2
        class="font-primary text-pretty text-2xl"
      >
        Categories
      </h2>
      <ul
        class="text-sm flex flex-col divide-y"
      >
        <li
          class="py-2 font-bold"
        >
          <a
            href="/category/arts-and-entertainment"
            title="Arts and Entertainment"
          >
            Arts and Entertainment
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/business-and-technology"
            title="Business and Technology"
          >
            Business and Technology
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/graphic-novels"
            title="Graphic Novels"
          >
            Graphic Novels
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/history-biography-and-social-science"
            title="History, Biography, and Social Science"
          >
            History, Biography, and Social Science
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/home-and-lifestyle"
            title="Home and Lifestyle"
          >
            Home and Lifestyle
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/horror"
            title="Horror"
          >
            Horror
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/languages-and-travel"
            title="Languages and Travel"
          >
            Languages and Travel
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/lgbtqia"
            title="LGBTQIA+"
          >
            LGBTQIA+
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/literature-and-poetry"
            title="Literature and Poetry"
          >
            Literature and Poetry
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/mystery"
            title="Mystery"
          >
            Mystery
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/romance"
            title="Romance"
          >
            Romance
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/science-and-math"
            title="Science and Math"
          >
            Science and Math
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/science-fiction-and-fantasy"
            title="Science Fiction and Fantasy"
          >
            Science Fiction and Fantasy
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/spirituality-and-philosophy"
            title="Spirituality and Philosophy"
          >
            Spirituality and Philosophy
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/wellness-sports-and-outdoors"
            title="Wellness, Sports, and Outdoors"
          >
            Wellness, Sports, and Outdoors
          </a>
        </li>
        <li
          class="py-2 "
        >
          <a
            href="/category/young-adult"
            title="Young Adult"
          >
            Young Adult
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
